src/Roslyn/Conventional.Roslyn/Conventions/SolutionDiagnosticAnalyzerConventionSpecification.cs

```diff
@@ -10,26 +12,43 @@ public interface ISolutionDiagnosticAnalyzerConventionSpecification
         IEnumerable<ConventionResult> IsSatisfiedBy(Solution solution);
     }
 
-    public abstract class SolutionDiagnosticAnalyzerConventionSpecification :
-        ISolutionDiagnosticAnalyzerConventionSpecification
+    public abstract class SolutionDiagnosticAnalyzerConventionSpecification : ISolutionDiagnosticAnalyzerConventionSpecification
     {
         // ReSharper disable once UnusedMemberInSuper.Global
         protected abstract string FailureMessage { get; }
+        protected readonly ConventionalSyntaxNodeAnalyzer Analyzer;
         private readonly string[] _fileExemptions;
 
-        protected SolutionDiagnosticAnalyzerConventionSpecification(string[] fileExemptions)
+        protected SolutionDiagnosticAnalyzerConventionSpecification(ConventionalSyntaxNodeAnalyzer diagnosticAnalyzer, string[] fileExemptions)
         {
+            Analyzer = diagnosticAnalyzer;
             _fileExemptions = fileExemptions;
         }
 
         public IEnumerable<ConventionResult> IsSatisfiedBy(Solution solution)
         {
-            return solution.Projects.SelectMany(x => x.Documents
-                    .Where(d => d.SupportsSyntaxTree)
-                    .Where(d => !_fileExemptions.Any(d.FilePath.EndsWith)))
-                .SelectMany(IsSatisfiedBy);
+            if (Analyzer.EnableConcurrentExecution())
+            {
+                var tasks = solution
```

Comment:
I'd suggest we use `Parallel.ForEach` here to limit how wide we go with task creation.